plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp

```diff
@@ -2011,6 +2032,48 @@ void CudaCalcAmoebaMultipoleForceKernel::getInducedDipoles(ContextImpl& context,
     }
 }
 
+
+void CudaCalcAmoebaMultipoleForceKernel::getTotalDipoles(ContextImpl& context, vector<Vec3>& dipoles) {
+    ensureMultipolesValid(context);
+    int numParticles = cu.getNumAtoms();
+    dipoles.resize(numParticles);
+    const vector<int>& order = cu.getAtomIndex();
+    if (cu.getUseDoublePrecision()) {
+        vector<double4> posqVec;
+        vector<double> labDipoleVec;
+        vector<double> inducedDipoleVec;
+        double totalDipoleVecX;
+        double totalDipoleVecY;
+        double totalDipoleVecZ;
+        inducedDipole->download(inducedDipoleVec);
+        labFrameDipoles->download(labDipoleVec);
+        cu.getPosq().download(posqVec);
+        for (int i = 0; i < numParticles; i++) {
+            totalDipoleVecX = posqVec[i].x * posqVec[i].w + labDipoleVec[3*i] + inducedDipoleVec[3*i];
+            totalDipoleVecY = posqVec[i].y * posqVec[i].w + labDipoleVec[3*i+1] + inducedDipoleVec[3*i+1];
+            totalDipoleVecZ = posqVec[i].z * posqVec[i].w + labDipoleVec[3*i+2] + inducedDipoleVec[3*i+2];
```

Comment:
I think that makes more intuitive sense. I was only thinking from the point
of view of the system I was dealing with. I'll remove the monopole
contribution and push the modified code then.

Thanks.

Saurabh

On 5 July 2016 at 11:45, peastman <notifications@github.com> wrote:

> In plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp
> <https://github.com/pandegroup/openmm/pull/1534#discussion_r69616583>:
>
> > +    dipoles.resize(numParticles);
> > +    const vector<int>& order = cu.getAtomIndex();
> > +    if (cu.getUseDoublePrecision()) {
> > +        vector<double4> posqVec;
> > +        vector<double> labDipoleVec;
> > +        vector<double> inducedDipoleVec;
> > +        double totalDipoleVecX;
> > +        double totalDipoleVecY;
> > +        double totalDipoleVecZ;
> > +        inducedDipole->download(inducedDipoleVec);
> > +        labFrameDipoles->download(labDipoleVec);
> > +        cu.getPosq().download(posqVec);
> > +        for (int i = 0; i < numParticles; i++) {
> > +            totalDipoleVecX = posqVec[i].x * posqVec[i].w + labDipoleVec[3*i] + inducedDipoleVec[3*i];
> > +            totalDipoleVecY = posqVec[i].y * posqVec[i].w + labDipoleVec[3*i+1] + inducedDipoleVec[3*i+1];
> > +            totalDipoleVecZ = posqVec[i].z * posqVec[i].w + labDipoleVec[3*i+2] + inducedDipoleVec[3*i+2];
>
> I think that would be more intuitive behavior. For example, for a
> non-polarizable point charge I'd expect it to return 0 for the dipole
> moment.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandegroup/openmm/pull/1534/files/a62b6bff8ad247678dd74eadfb749c63bb896bcb#r69616583>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ANPF-2Q_YvaEQge-3w3pVlG8jbWaurArks5qSqZVgaJpZM4JCpXy>
> .
>